browser/scripts/ui-core.js

```diff
@@ -417,117 +498,28 @@ VizorUI.prototype.onFullScreenChanged = function() {	// placeholder
-VizorUI.prototype.applyVisibility = function(andUpdateState) {
-	andUpdateState = (typeof andUpdateState === 'undefined') ? true : !!andUpdateState;
-	var dom = this.dom;		// normally E2.dom
-	var state = this.state;
-	var visibility = state.visibility;
-	var $assets = dom.assetsLib, $presets = dom.presetsLib, $chat = dom.chatWindow;
-	var $patch_editor = dom.canvas_parent;
-	var show_panels = state.visible && visibility.floating_panels;
-
-	$assets.toggle(show_panels && visibility.panel_assets)
-		.toggleClass('uiopen', show_panels && visibility.panel_assets);
-	$presets.toggle(show_panels && visibility.panel_presets)
-		.toggleClass('uiopen', show_panels && visibility.panel_presets);
-	$chat.toggle(show_panels && visibility.panel_chat)
-		.toggleClass('uiopen', show_panels && visibility.panel_chat);
-
-	dom.btnAssets.toggleClass('ui_off', !visibility.panel_assets);
-	dom.btnPresets.toggleClass('ui_off', !visibility.panel_presets);
-	dom.btnChatDisplay.toggleClass('ui_off', !visibility.panel_chat);
-	dom.btnGraph.toggleClass('ui_off', !visibility.patch_editor);
-
-	$patch_editor.toggle(state.visible && visibility.patch_editor);
-	if (andUpdateState) this.updateState();
-	this.enforceConstraints();
-
-	var inBuildMode = !visibility.patch_editor,
-		inProgramMode = visibility.patch_editor;
-
-	// sync camera buttons
-	var worldEditorActive = state.viewCamera === uiViewCam.world_editor;
-
-
-	dom.btnBuildMode.toggleClass('ui_on', inBuildMode)
-		.toggleClass('ui_off', inProgramMode);
-	dom.btnProgramMode.toggleClass('ui_on', inProgramMode)
-		.toggleClass('ui_off', inBuildMode);
-
-	dom.btnSavePatch.attr('disabled', inBuildMode);
-	dom.btnInspector.attr('disabled', inBuildMode);
-
-//  these will still zoom, whatever the mode?
-	/*
-	dom.btnZoomOut.attr('disabled',isProgramMode);
-	dom.btnZoom.attr('disabled',worldEditorActive);
-	dom.btnZoomIn.attr('disabled',worldEditorActive);
-	*/
-
-	dom.btnMove.attr('disabled',inProgramMode);
-	dom.btnScale.attr('disabled',inProgramMode);
-	dom.btnRotate.attr('disabled',inProgramMode);
-	dom.btnEditorCam.parent().toggleClass('active', worldEditorActive);
-	dom.btnVRCam.parent().toggleClass('active', !worldEditorActive);
-
-	if (inProgramMode) {
-		dom.tabPresets.find('a').trigger('click');
-	} else {
-		dom.tabObjects.find('a').trigger('click');
-		dom.tabPresets.addClass('inactive ui_off');
-	}
-
-	E2.app.noodlesVisible = inProgramMode;
-};
-
-
 /***** TOGGLE LAYERS OF THE UI ON OR OFF *****/
 VizorUI.prototype.toggleFloatingPanels = function(forceVisibility) {
 	var v = this.state.visibility;
-	if (typeof forceVisibility !== 'undefined')
-		v.floating_panels = forceVisibility;
-	else
-		v.floating_panels = !v.floating_panels;
-
-	this.applyVisibility();
+	if ((forceVisibility !== false) && (!v.panel_assets) && (!v.panel_chat) && (!v.panel_presets)) {
+		// v._.panel_assets = true;
+		v._.panel_chat = true;
```

Comment:
visibility.floating panels and .visible are not simply shorthand methods, modifying both generates slave events. in special cases such as this (e.g. the panels are visible but each has been switched off) and similar, to avoid multiple triggering of the same event, the state is modified directly.